### ISSUE TITLE
DB Creation Errors

### ISSUE THREAD

**Issue by Kolano** [cuckoosandbox/cuckoo#1516]
Seeing the following error during initial Cuckoo-web startup / DB creation /w MySQL DB...

```
Traceback (most recent call last):
  File "/home/pass/cuckoo_ve/bin/cuckoo", line 11, in <module>
    load_entry_point('Cuckoo==2.0.2', 'console_scripts', 'cuckoo')()
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/cuckoo/main.py", line 488, in web
    Database().connect()
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/cuckoo/core/database.py", line 446, in connect
    self._create_tables()
  File "/home/pass/cuckoo_ve/local/lib/python2.7/site-packages/cuckoo/core/database.py", line 454, in _create_tables
    "Unable to create or connect to database: %s" % e
cuckoo.common.exceptions.CuckooDatabaseError: Unable to create or connect to database: (_mysql_exceptions.OperationalError) (1050, "Table 'errors' already exists") [SQL: u'\nCREATE TABLE errors (\n\tid INTEGER NOT NULL AUTO_INCREMENT, \n\taction VARCHAR(64), \n\tmessage TEXT NOT NULL, \n\ttask_id INTEGER NOT NULL, \n\tPRIMARY KEY (id), \n\tFOREIGN KEY(task_id) REFERENCES tasks (id)\n)\n\n']
```

Seems to start up correctly post such...

```
Performing system checks...

System check identified no issues (0 silenced).
May 02, 2017 - 21:23:16
Django version 1.8.4, using settings 'cuckoo.web.web.settings'
Starting development server at http://0.0.0.0:8000/
Quit the server with CONTROL-C.
```

**Comment by Kolano** [comment]
This is with Cuckoo 2.0.2

**Comment by jbremer** [comment]
Did you do anything special to the database before starting Cuckoo? Do other applications also access it? Could you show a listing of all currently available tables?
Can you reproduce this (by using a new/clean database)? Thanks!

**Comment by Kolano** [comment]
- Nothing special is done to the DB, using the apt-get install and mysql_secure_installation script.
- Other apps can access the DB server, as can Cuckoo. It creates the appropriate tables, not clear why it's trying to do so again.
- ![cuckoo tables](https://cloud.githubusercontent.com/assets/6509657/25646312/2fba15b8-2f85-11e7-88f9-24599c7934a0.png)
- Wasn't able to reproduce by dropping the DB, re-establishing it, and restarting; will see if it re-occurs generally with my next deployment.

**Comment by RANGERBEE** [comment]
What I did to "cure" DB errors was the following:

Run clean
Run migrate
Run clean
Start cuckoo

**Comment by Kolano** [comment]
It's not clear those commands would have any impact when the DB doesn't yet exist. I'm seeing this with a fresh deploy to an environment without a pre-existing DB.

I will be reattempting a deploy /w 2.0.3a3 later today, will report back regarding if I can reproduce it with such.

**Comment by jbremer** [comment]
I'm somewhat considering a race condition in `cuckoo web` for this issue. The Django HTTP server always creates an additional process (the actual worker) aside from the parent process. Due to this two database connections will be made. It could be possible these somehow interfere with each other due to missing database locks or whatever.
What happens if you `cuckoo -d` first before running the `cuckoo web` commands? Thanks!

**Comment by Kolano** [comment]
I'm using Supervisor to run Cuckoo, which it is started just prior to cuckoo-web. so perhaps it is some conflict between their 2 startups. I had presumed cuckoo-web was responsible for the DB setup.

**Comment by jbremer** [comment]
Whichever is the first to find the completely empty database will attempt to create tables. Provided this is done through transactions and there might be some small delays etc.. best advice I can give is - actually run Cuckoo once or twice before deploying everything completely.

**Comment by jbremer** [comment]
An internal delay would only make the startup times slower and/or less predictable. A flag would have to be a file created on the file system, but whether that's such a clean solution.. ;-)
I've looked into that (for different reasons) in the past and you'd probably end up with the `lockfile` package (if I remember correctly) which, well, meh. Keep in mind that this would have to work on Windows, Mac OS X, and Linux.

For reasons stated above I don't think we're going to be adding such a feature. In the future, just run the `cuckoo` command once manually. You're supposed to do that before invoking the other subcommands anyway. Closing as "bug resolved" "wontfix". If you have alternative suggestions as to why this feature should be reconsidered, feel free to reopen / etc to let us know!
